src/protocol/dot_product.rs

```diff
@@ -0,0 +1,162 @@
+/// "High-Throughput Secure AES Computation" (High-throughput secure AES computation. In WAHC@CCS 2018, pp. 13–24, 2018)
+/// by Koji Chida, Koki Hamada, Dai Ikarashi, Ryo Kikuchi and Benny Pinkas
+/// <https://dl.acm.org/doi/10.1145/3267973.3267977>
+///
+/// The extension allows for the computation of the sum of multiple multiplications (in other words, a dot-product operation)
+/// at the communication cost of a single multiplication.
+///
+/// This extension is summarized in the paper:
+/// "Fast Large-Scale Honest-Majority MPC for Malicious Adversaries"
+/// by Koji Chida, Daniel Genkin, Koki Hamada, Dai Ikarashi, Ryo Kikuchi, Yehuda Lindell, and Ariel Nof
+/// <https://link.springer.com/content/pdf/10.1007/978-3-319-96878-0_2.pdf>
+///
+/// In that paper, this functionality is called `F_product` and
+/// an instantiation for Replciated Secret Sharing is provided in section 6.1 on page 27
+///
+/// To summarize the CHIKP protocol:
+/// each helper computes `u_i+1 = (s_i + s_i+1) * (t_i + t_i+1) - s_i+1 * t_i+1`
+/// It then randomizes this share with correlated randomness, and sends it to the helper to its right,
+/// and receives `u_i` from the helper to its left.
+/// Each helper then defines the pair `(u_i, u_i+1)` as its shares of the output.
+///
+/// (Note it is equivalent to compute `u_i+1 = s_i * t_i + s_i+1 * t_i + s_i * t_i+1`,
+/// but this requires three multiplications instead of two. Assuming multiplying field elements is
+/// significantly more costly (in CPU terms) than addition, this is less efficient to compute)
+///
+/// The key observation, that makes this protocol work, is that each helper can compute many `u_i+1` values
+/// from many multiplications, sum them all together, randomize them with correlated randomness, and can
+/// send this single field value to the next helper. The result will be a replicated secret sharing of the
+/// dot-product of the two vectors.
+///
+/// ## Errors
+/// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+/// back via the error response
+#[allow(dead_code)]
+pub async fn dot_product<F: Field, N: Network>(
+    ctx: ProtocolContext<'_, N>,
+    record_id: RecordId,
+    a: &[Replicated<F>],
+    b: &[Replicated<F>],
+) -> Result<Replicated<F>, BoxError> {
+    assert_eq!(a.len(), b.len(), "lengths of the lists must be the same");
+
+    let prss = &ctx.prss();
+    let (s_left, s_right): (F, F) = prss.generate_fields(record_id.into());
+
+    let mut u_right = s_right - s_left;
+    for i in 0..a.len() {
+        let (a_left, a_right) = a[i].as_tuple();
+        let (b_left, b_right) = b[i].as_tuple();
+        u_right += (a_left + a_right) * (b_left + b_right) - a_right * b_right;
+    }
+
+    // send our `u_i+1` value to the helper on the right
+    let channel = ctx.mesh();
+    channel
+        .send(
+            channel.identity().peer(Direction::Right),
+            record_id,
+            UValue { u: u_right },
```

Comment:
This bit where we wrap values in a new type is not ergonomic.  We should just be able to send the field value.

Reply:
I agree. We should make that change. @akoshelev - is there a specific reason why we are doing this? Can we just stop using these structs?

Reply:
the original idea was to use the type information to resolve the ambiguity in cases where protocols may want to send two or more messages in the same step (and record). Infra theoretically can support this by resolving `record_id` to multiple messages. Having said that I am yet to see a use case for it, so I don't have a good argument against getting rid of wrappers.